`#assert` parsing, typchecking, and SILGen (hidden under flag)

marcrasi opened the pull request:
Part 1 of mega-patch https://github.com/apple/swift/pull/19579.

One possible issue is that the flag doesn't disable the stuff in `utils/gyb_syntax_support/ExprNodes.py` and `utils/gyb_syntax_support/Token.py`, so if there's anything that picks up on that (e.g. code completion tools?), it won't respect the flag. Is this an issue? Anyone have suggestions on how to fix it?

harlanhaskins commented on lib/SILOptimizer/Mandatory/IRGenPrepare.cpp:
```diff
@@ -41,14 +42,15 @@ static bool cleanFunction(SILFunction &fn) {
       ++i;
 
-      // Remove calls to Builtin.staticReport().
+      // Remove calls to Builtin.staticReport() and Builtin.staticReport().
```
`Builtin.poundAssert()`

marcrasi replied:
yep, thanks, fixed!

ravikandhadai commented:
LGTM

xedin reviewed (approved):
Sema related things look good to me!

rintaro commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3507,3 +3511,63 @@ Parser::parsePlatformVersionConstraintSpec() {
+    message = cast<StringLiteralExpr>(messageExpr)->getValue();
+  }
+
+  if (parseToken(tok::r_paren, endLoc, diag::pound_assert_expected, ")")) {
```
We usually use `parseMatchingToken()` for this.

marcrasi replied:
done

rintaro commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3507,3 +3511,63 @@ Parser::parsePlatformVersionConstraintSpec() {
+    return errorAndSkipToEnd();
+  }
+
+  StringRef message = "assertion failed";
```
I don't think Parser should have the default message here. Could you make the default `StringRef()` and handle it in the actual assertion error logic?

rintaro commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -1697,6 +1697,10 @@ ParserResult<Expr> Parser::parseExprPrimary(Diag<> ID, bool isExprBasic) {
 
+  case tok::pound_assert:
+    return parseExprPoundAssert();
+    break;
```
`#assert()` should really be modeled as a expression? From Sema code, I understand the type of `#assert(...)` expression is always `Void`. But I don't think `let foo: Void = #assert(true)` should even be accepted.
Why don't you model it as a statement?

rintaro commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3507,3 +3511,63 @@ Parser::parsePlatformVersionConstraintSpec() {
+    return errorAndSkipToEnd();
+  }
+
+  auto conditionExprResult = parseExpr(diag::pound_assert_expected_expression);
```
If `parseExpr()` has code completion status. We have to propagate it to the caller. Could you:
```cpp
if (conditionExprResult.hasCodeCompletion())
  return conditionExprResult;
```

rintaro commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3507,3 +3511,63 @@ Parser::parsePlatformVersionConstraintSpec() {
+      endLoc = PreviousLoc;
+    }
+    return makeParserResult<Expr>(new (Context)
+                                      ErrorExpr(SourceRange(startLoc, endLoc)));
```
This is nice, but I don't think every expression parsing should have recovery logic on its own. We usually let `parseExprList()` or `parseBraceItems()` to recover from parse error.
Any particular reason to recover here? If there's no reason, diagnose and return `makeParserError()` should be sufficient.

rintaro replied:
No. Returning error makes `parseBraceItems()` [automatically recover](https://github.com/apple/swift/blob/b01776164569cece2706d048707be37e970ce585/lib/Parse/ParseStmt.cpp#L468-L477). You don't need to touch `parseBraceItems()`.

rintaro commented:
> One possible issue is that the flag doesn't disable the stuff in utils/gyb_syntax_support/ExprNodes.py and utils/gyb_syntax_support/Token.py, so if there's anything that picks up on that (e.g. code completion tools?), it won't respect the flag. Is this an issue? 

Currently there's no runtime/compile-time mechanism to enable/disable `libSyntax` node depending on flags. But I don't think that's a big issue. At least, we don't have such source tools which automatically lists up declared node kinds.

nkcsgexi commented on utils/gyb_syntax_support/ExprNodes.py:
```diff
@@ -569,4 +569,19 @@
+             Child('Message', kind='StringLiteralExpr', is_optional=True,
+                   description='The assertion message.'),
+             Child('RightParen', kind='RightParenToken'),
+         ]),
```
Could you add some #assert tests in round_trip_parse_gen.swift to ensure they are constructed as syntax nodes correctly?

nkcsgexi replied:
Missing tags indicates unknown syntax kinds, which unfortunately means these syntax nodes are not constructed properly. You can use command `$./swift-macosx-x86_64/bin/swift-syntax-test -input-source-filename /tmp/t.swift -parse-gen -print-node-kind -print-trivial-node-kind` to print all unknown kinds and diagnose what's wrong with it.

marcrasi replied:
Thanks, this helped me figure it out. The problem was that the test runs without "enable-experimental-pound-assert" so the parser bails out before parsing the whole statement, which confuses the syntax thing. I fixed it by making the parser parse the whole statement before bailing out.

nkcsgexi commented on lib/Sema/CSGen.cpp:
```diff
@@ -3102,6 +3102,18 @@ namespace {
+
+      // #assert(...) has type Void.
+      return CS.getASTContext().TheEmptyTupleType;
+    }
```
Can we use `TypeChecker::typeCheckStmtCondition` to type check the pound condition just like for other statements if we module #assert as a statement instead of an expression like @rintaro suggested above? Generating constraint conditions seems to me an overkill for this case. cc: @xedin

xedin replied:
I agree, it seems like we are to go statement route, it shouldn't require any type-checker changes, but should use existing `tyeCheckStmtCondition` mechanism.

marcrasi commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -3507,3 +3511,63 @@ Parser::parsePlatformVersionConstraintSpec() {
+      return errorAndSkipToEnd();
+    }
+
+    message = cast<StringLiteralExpr>(messageExpr)->getValue();
```
I refactored it a bit and started using it.

marcrasi reviewed (commented):
Thanks for all the comments! I addressed most of them, but I have 3 questions inline:
https://github.com/apple/swift/pull/19618#discussion_r228025801
https://github.com/apple/swift/pull/19618#discussion_r228025959
https://github.com/apple/swift/pull/19618#discussion_r228026185

rintaro commented on lib/Parse/ParseStmt.cpp:
```diff
@@ -2400,3 +2405,54 @@ ParserResult<CaseStmt> Parser::parseStmtCase(bool IsActive) {
+  // if (conditionExprResult.hasCodeCompletion())
+  //   return conditionExprResult;
+  if (conditionExprResult.isParseError())
+    return makeParserError();
```
`return ParserStatus(conditionExprResult)` should be OK.

marcrasi replied:
thanks, done!

rintaro commented on lib/Parse/ParseStmt.cpp:
```diff
@@ -2400,3 +2405,54 @@ ParserResult<CaseStmt> Parser::parseStmtCase(bool IsActive) {
+///   '#assert' '(' expr (',' string_literal)? ')'
+ParserResult<Stmt> Parser::parseStmtPoundAssert() {
+  SyntaxParsingContext AssertContext(SyntaxContext,
+      SyntaxKind::PoundAssertStmt);
```
Statement libSyntax parsing is a bit different from expression parsing. The caller (`parseStmt()`) have already [set up the context for the statement](https://github.com/apple/swift/blob/b01776164569cece2706d048707be37e970ce585/lib/Parse/ParseStmt.cpp#L547). So you should just set the syntax kind of it:
```cpp
SyntaxContext->setCreateSyntax(SyntaxKind::PoundAssertStmt);
```

marcrasi replied:
thanks, done!

marcrasi commented:
Thanks for the answers! I've addressed the additional comments, and I don't have any questions left.

rintaro commented on utils/gyb_syntax_support/StmtNodes.py:
```diff
@@ -326,4 +326,18 @@
+                   description='The assertion condition.'),
+             Child('Comma', kind='CommaToken', is_optional=True,
+                   description='The comma after the assertion condition.'),
+             Child('Message', kind='StringLiteralExpr', is_optional=True,
```
`StringLiteralExpr ` -> `StringLiteralToken`.
This is why the last test case in `test/Syntax/round_trip_parse_gen.swift` isn't wrapped with `<PoundAssertStmt />`

marcrasi replied:
oops, didn't notice that. thanks for catching. fixed!

rintaro commented:
Parser changes LGTM otherwise!

ravikandhadai commented:
@swift-ci please smoke test

ravikandhadai commented:
@swift-ci Please smoke test

marcrasi commented:
These failures look unrelated to what I did. I'll try merging with the latest master and running them again, to see what happens.

marcrasi commented:
@swift-ci Please smoke test

marcrasi commented:
@swift-ci Please smoke test

marcrasi commented:
@swift-ci Please smoke test

marcrasi commented:
@swift-ci Please smoke test

ravikandhadai commented:
Yeah the failures seem related to the CI.

marcrasi commented:
What's the procedure now? Do we need any more approvals? Should I clean up the commit history before we merge?

devincoughlin commented:
Yes, it would be good to clean up the commit history before you merge and add an informative commit message for the entire commit.

Commit:
parsing, typechecking, and SILGen for `#assert`

`#assert` is a new static assertion statement that will let us write
tests for the new constant evaluation infrastructure that we are working
on. `#assert` works by lowering to a `Builtin.poundAssert` SIL
instruction. The constant evaluation infrastructure will look for these
SIL instructions, const-evaluate their conditions, and emit errors if
the conditions are non-constant or false.

This commit implements parsing, typechecking and SILGen for `#assert`.

marcrasi commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -1249,6 +1254,19 @@ ERROR(expr_typeof_expected_rparen,PointsToFirstBadToken,
+      "expected a condition expression", ())
+ERROR(pound_assert_expected_string_literal,PointsToFirstBadToken,
+      "expected a string literal", ())
+ERROR(pound_assert_string_interpolation,PointsToFirstBadToken,
```
done

marcrasi commented on lib/SILGen/SILGenStmt.cpp:
```diff
@@ -507,6 +507,26 @@ void StmtEmitter::visitYieldStmt(YieldStmt *S) {
   SGF.emitYield(S, sources, origTypes, SGF.CoroutineUnwindDest);
 }
 
+void StmtEmitter::visitPoundAssertStmt(PoundAssertStmt *S) {
```
Oh yep, I found some tests in our branch that I forgot to pull into this patch. I have pulled them in.

marcrasi commented:
I addressed @devincoughlin's comments and cleaned up the commit history.

gottesmm commented on test/SILGen/pound_assert.swift:
```diff
@@ -0,0 +1,19 @@
+  // CHECK: [[GET_LOGIC_VALUE:%.*]] = function_ref {{.*}}_getBuiltinLogicValue
+  // CHECK-NEXT: [[LOGIC_VALUE:%.*]] = apply [[GET_LOGIC_VALUE]]
+  // CHECK-NEXT: [[MESSAGE:%.*]] = string_literal utf8 "custom message"
+  // CHECK-NEXT: builtin "poundAssert"([[LOGIC_VALUE]] : $Builtin.Int1, [[MESSAGE]] : $Builtin.RawPointer)
```
One suggestion. SIL emits the function name at the end brace of a sil function so you can pattern match the end of a function uniquely. Can you do that here? E.g.:

```
// CHECK: } // end sil function '$s12pound_assert13customMessage{{[_0-9a-zA-Z]*}}'
```

For both of the tests.

gottesmm commented on lib/SILGen/SILGenStmt.cpp:
```diff
@@ -507,6 +507,26 @@ void StmtEmitter::visitYieldStmt(YieldStmt *S) {
   SGF.emitYield(S, sources, origTypes, SGF.CoroutineUnwindDest);
 }
 
+void StmtEmitter::visitPoundAssertStmt(PoundAssertStmt *S) {
```
Can you change this to be a lower case s? (or even better stmt).

gottesmm reviewed (approved):
LGTM with my nits.

ravikandhadai commented:
@marcrasi We have all the reviews. I think we are good to test and merge this, once the latest comments are fixed.

marcrasi commented:
I addressed @gottesmm's comments and rebased to master.

I had to add a switch case to `lib/SILOptimizer/Transforms/AccessEnforcementReleaseSinking.cpp` for the rebase.

I'm not familiar with the process for running final tests and merging to master. @ravikandhadai, could you tell me how to do it, or just do it yourself? Thanks!

ravikandhadai commented:
@swift-ci Please smoke test

ravikandhadai commented:
@swift-ci Please smoke test

ravikandhadai commented:
@marcrasi There is a new merge conflict, but it looks trivial to fix.

marcrasi commented:
@swift-ci Please smoke test

marcrasi commented:
rebased and fixed the conflict

ravikandhadai commented:
@swift-ci please test benchmark

ravikandhadai commented:
@swift-ci Please smoke benchmark

ravikandhadai commented:
@swift-ci Please smoke benchmark

ravikandhadai commented:
@swift-ci Please smoke test Linux platform

ravikandhadai commented:
@swift-ci Please smoke test Linux platform

ravikandhadai commented:
@swift-ci Please test and merge

ravikandhadai commented:
@swift-ci Please test and merge

ravikandhadai commented:
@swift-ci Please test Linux platform

ravikandhadai commented:
@swift-ci Please test Linux Platform

ravikandhadai commented:
@marcrasi Linux builds are failing repeatedly. They are running out of time. I am trying figure out what the issue could be.

shahmishal commented:
@swift-ci test Linux

marcrasi commented:
I see. Looks like it's timing out while running lldb tests? I'll see if I can reproduce on my linux machine.

ravikandhadai commented:
@marcrasi It could be a problem/flakiness with the CI. @shahmishal has increased the timeout  and started the test again. Let us wait for this test.

marcrasi commented:
Ok! Sounds good.

ravikandhadai commented:
@swift-ci Please test and merge
